Improved opt out

bx80 opened the pull request:
### Description:

Fixes #17452 

This is PR contains the UI changes to allow a choice between the tracker opt-out and the self-contained opt-out. 

The self-contained code is ~85 lines and this contains a small 'MatomoConsent' class to handle basic checking, reading and writing of the consent cookie. All settings, including single language translations, are passed in as an object so they are easily editable at the top of the code block.

The 'MatomoConsent' class is reused the fallback method for the tracker opt out JavaScript to set cookies directly should the Matomo tracker be unavailable. 

The tracker opt-out supports the following URL parameters and is compatible with the existing iFrame opt-out URL (just change `optOut` to `optOutJS` and add the div) : 
```
backgroundColor
fontColor
fontSize
fontFamily
language              default "auto"            Language code for translations ("auto" for browser language)
showIntro             default 1                 Should the opt-out intro text be shown?
divId                 default "matomo-opt-out"  The id of the div which will contain the opt-out form
useCookiesIfNoTracker default 1                 Directly read/written consent cookies if no tracker found?
useCookiesTimeout     default 10                How long to wait for the tracker to be detected?
useSecureCookies      default 1                 Set secure cookies?
```

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

Commit:
New opt out UI, self-contained code - WIP

sgiehl commented on plugins/PrivacyManager/lang/en.json:
```diff
@@ -161,13 +161,17 @@
         "UseAnonymizeUserId": "Anonymize User ID",
         "PseudonymizeUserId": "Replace User ID with a pseudonym",
-        "OptOutCustomize": "Customize the Opt-out iframe",
+        "OptOutCustomize": "Customize the opt-out appearance",
```
You should use a new translation key here. The meaning of the content changed too much. Existing translations for other languages would otherwise still be used and might cause confusion.

bx80 replied:
Good spot, thanks :slightly_smiling_face:

Commit:
Updated translation

Commit:
Merge branch '4.x-dev' into m-17452-opt-out-first-party

bx80 committed:
built vue files

Commit:
Added OptOutJS functionality, code improvements, show custom guide link on UI

Commit:
Merge remote-tracking branch 'origin/m-17452-opt-out-first-party' into m-17452-opt-out-first-party

# Conflicts:
#	plugins/PrivacyManager/vue/dist/PrivacyManager.umd.js
#	plugins/PrivacyManager/vue/dist/PrivacyManager.umd.min.js

Commit:
Merge branch '4.x-dev' into m-17452-opt-out-first-party

Commit:
Rebuild Vue

bx80 commented:
@sgiehl I've haven't implemented the requirement that if the tracker cannot be found then detect if the user is logged into the Matomo dashboard and show an error on the webpage and log an application error. I'm not seeing a straightforward way to do this without trying to read Matomo session cookies and add new API calls to log errors, which seems like an overkill. Any suggestions?

bx80 committed:
built vue files

Commit:
Fix typo

Commit:
Merge remote-tracking branch 'origin/m-17452-opt-out-first-party' into m-17452-opt-out-first-party

Commit:
Default iframe out-out to show intro, minor code tidy, updated UI test screenshots

bx80 commented:
In case it's helpful for testing, this is the tracking page I used during development, it can load the tracker after a random delay (or not at all) and includes a URL string with all the parameters.
```
<html>
  <head>
    <meta charset="utf-8">
    <title>OptOutJS</title>

    <!-- Matomo -->
    <script>

    // Test variables
    let loadTracker = true;
    let maxRandomDelay = 10;
    let minRandomDelay = 5;
    // ---

    function loadMatomoTracker() {
      console.log('.. now loading Matomo tracker');
      var _paq = window._paq = window._paq || [];
      _paq.push(['trackPageView']);
      _paq.push(['enableLinkTracking']);
      (function() {
        var u="//matomo/";
        _paq.push(['setTrackerUrl', u+'matomo.php']);
        _paq.push(['setSiteId', '3']);
        var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
        g.async=true; g.src=u+'matomo.js'; s.parentNode.insertBefore(g,s);
      })();
    }

    if (loadTracker === true)  {
      let randomInterval = (Math.floor((Math.floor(Math.random() * ((maxRandomDelay * 1000) - (minRandomDelay * 1000)) ) + (minRandomDelay * 1000)) / 1000) * 1000);
      setTimeout(function() { loadMatomoTracker() }, randomInterval);
      console.log('delaying Matomo tracker load for '+randomInterval / 1000+' seconds...');
    } else {
      console.log('tracker loading disabled');
    }
    </script>
    <!-- End Matomo Code -->

  </head>
  <body>
  OptOutJS
  <br><br>

  <div id="matomo-opt-out" style="text-align:center;width:350px;padding:10px">...opt-out loading...</div>
  <script src="https://matomo//index.php?module=CoreAdminHome&action=optOutJS&useCookiesTimeout=10&div=m-opt-out&language=auto&backgroundColor=BBBBFF&fontColor=444444&fontSize=16px&fontFamily=TimesNewRoman&showIntro=1&useCookiesIfNoTracker=1"></script>

  </body>
</html>
```

sgiehl commented on plugins/CoreAdminHome/OptOutManager.php:
```diff
@@ -158,10 +181,324 @@ public function getQueryParameters()
+$code = <<<HTML
+<div id="matomo-opt-out" style=""></div>
+<script>    
+    let settings = {};         
```
I guess we should use `var` instead of `let`, to ensure this would also work on older browsers.

sgiehl commented:
> I've haven't implemented the requirement that if the tracker cannot be found then detect if the user is logged into the Matomo dashboard and show an error on the webpage and log an application error. I'm not seeing a straightforward way to do this without trying to read Matomo session cookies and add new API calls to log errors, which seems like an overkill. Any suggestions?

I guess I would have tried to solve that using the opt out code. The page, that includes the optout, fetches some javascript from Matomo, so in case the user is logged in you could simply add some more javascript code, that if `matomo.js` can't be loaded displays a warning and/or sends back a request to matomo, to log an error. But that for sure would only work if the javascript returned for the optout wasn't cached.
Maybe @tsteur has a better idea.

Commit:
Merge branch '4.x-dev' into m-17452-opt-out-first-party

bx80 committed:
built vue files

bx80 commented:
> The page, that includes the optout, fetches some javascript from Matomo, so in case the user is logged in you could simply add some more javascript code, that if matomo.js can't be loaded displays a warning and/or sends back a request to matomo, to log an error.

That was my plan, but the optOutJS request in many cases is likely to be cross-site and therefore doesn't pass any Matomo session cookies (which are `Same-site:Lax` and `httpOnly` by default), so when processing the optOutJS request Matomo doesn't know about the current session and is unaware that the user is logged in. For the similar reasons there's not going to be a secure way for the optOutJS code running on `domain a` to access the Matomo session for `domain b` at runtime either.

Without access to the Matomo API I suppose we could still use the tracking API to force some sort of tracking failure log (eg. `GET https://matomo/matomo.php?idsite=9999998&rec=1`), but I'm not sure that we'd want to create failure logs for every visitor who can't load the tracker JS.

Commit:
Replaced let with var

Commit:
Merge remote-tracking branch 'origin/m-17452-opt-out-first-party' into m-17452-opt-out-first-party

sgiehl commented:
@bx80 Good points. But I actually don't have any further ideas how to solve that requirement.
Maybe we should discuss this requirement with @tsteur again. I understand the purpose why this could be helpful. But actually I'm wondering if there might be cases where it might be on purpose that the optout might be shown without having the tracker js file loaded.
For example a consent manager could provide opt outs for certain tracking mechanisms, but might prevent loading any tracking code before it's accepted or declined maybe. Not sure if a warning would annoy a Matomo user in that case.

tsteur commented:
> But I actually don't have any further ideas how to solve that requirement. Maybe we should discuss this requirement with @tsteur again. 

does this refer to below requirement?

> I've haven't implemented the requirement that if the tracker cannot be found then detect if the user is logged into the Matomo dashboard and show an error on the webpage and log an application error. I'm not seeing a straightforward way to do this without trying to read Matomo session cookies and add new API calls to log errors, which seems like an overkill. Any suggestions?

It could be fine to not implement as long as people have a chance to notice that the opt in won't work if the JS tracker file isn't loaded on the same page. It could otherwise cause legal issues if the opt out is embedded and not working.

Note: I believe when using a consent manager then the opt out is typically handled through the consent manager and not through an extra opt out screen unless they use both (which is maybe not done normally)

bx80 commented:
>It could be fine to not implement as long as people have a chance to notice that the opt in won't work if the JS tracker file isn't loaded on the same page. It could otherwise cause legal issues if the opt out is embedded and not working.

@tsteur By default we have a fallback method included which will directly set the required consent cookies if the JS tracker isn't loaded on the page. We're not currently providing a UI option to disable this. So unless someone deliberately adds the extra URL parameter to disable the fallback method then it there doesn't seem to be a way to have the opt-out fail silently.

Tracker? Use tracker to set consent.
No tracker? Set consent cookies directly (also console log)
No cookies? Show error on page.
No https? Show error on page.
No opt out div found? (console log only)

The weak point here is if the opt out div is not found on the page. It might be better to add a div to the page and show a visible error message instead of just a console log. If I make that change then perhaps we don't need to implement logged in user detection / application logs?

tsteur commented:
> No tracker? Set consent cookies directly (also console log)

This would only work if they have configured cookie settings somehow? Like it may cause issues if they otherwise use different cookie settings? I suppose there's no way for us to detect this. Maybe we can give people clear instructions on how to test that the opt out works?

> The weak point here is if the opt out div is not found on the page. It might be better to add a div to the page and show a visible error message instead of just a console log. If I make that change then perhaps we don't need to implement logged in user detection / application logs?

Sounds good. Note that the JS file may be loaded before the div is there depending on how the file is being loaded.

bx80 commented:
> > No tracker? Set consent cookies directly (also console log)
> 
> This would only work if they have configured cookie settings somehow? Like it may cause issues if they otherwise use different cookie settings? I suppose there's no way for us to detect this. Maybe we can give people clear instructions on how to test that the opt out works?

If they have configured custom cookie settings in Matomo (cookie_path, cookie_domain, etc) then those settings will be included in the served JavaScript and will be used, otherwise the default Matomo consent cookie settings will be used. It should always be using the same cookie settings as the JS tracker,

We could add an info box on the opt-out code screen encouraging testing of the opt-out. Something like:

---

**Remember to test your opt-out!**
To ensure that your opt-out is working properly with your website it is recommended to perform the following quick test after adding the opt out code.
- Browse to your main website page.
- Once the page has loaded you should see the opt-out box.
- Unchecking the checkbox to opt-out should show the opted-out message.
- Checking the checkbox to opt-in should show the opted-in message.

---

> 
> > The weak point here is if the opt out div is not found on the page. It might be better to add a div to the page and show a visible error message instead of just a console log. If I make that change then perhaps we don't need to implement logged in user detection / application logs?
> 
> Sounds good. Note that the JS file may be loaded before the div is there depending on how the file is being loaded.

The main opt out code initializes from a `DOMContentLoaded` event listener, so hopefully this will make sure the div has always had chance to be rendered before we check for it.

tsteur commented:
> If they have configured custom cookie settings in Matomo (cookie_path, cookie_domain, etc) then those settings will be included in the served JavaScript and will be used, otherwise the default Matomo consent cookie settings will be used. It should always be using the same cookie settings as the JS tracker,

FYI typically people don't configure cookie settings in Matomo as these settings only apply to third party cookie. The first party cookie is configured in the JS tracking code usually. Meaning the config in Matomo would likely not be of much help (unless I'm missing something which is possible).

> We could add an info box on the opt-out code screen encouraging testing of the opt-out. Something like:

That be great 👍 

> The main opt out code initializes from a DOMContentLoaded event listener, so hopefully this will make sure the div has always had chance to be rendered before we check for it.

Awesome 👍

Commit:
Merge branch '4.x-dev' into m-17452-opt-out-first-party

Commit:
Improved warning when opt-out content div is missing, added info box to recommend testing

bx80 committed:
built vue files

Commit:
Update UI test screenshots

Commit:
Merge remote-tracking branch 'origin/m-17452-opt-out-first-party' into m-17452-opt-out-first-party

Commit:
Merge branch '4.x-dev' into m-17452-opt-out-first-party

sgiehl commented on plugins/CoreAdminHome/templates/optOut.twig:
```diff
@@ -50,9 +50,11 @@
 
         {{ 'CoreAdminHome_OptOutCompleteBis'|translate }}
         </p>
+        {% if showIntro %}
```
Shouldn't that also hide the `opt-out complete` message? So there is literally only the checkbox? 🤔

sgiehl commented on plugins/CoreAdminHome/Controller.php:
```diff
@@ -255,11 +255,22 @@ public function trackingCodeGenerator()
+     * @return string
+     * @throws Exception
+     */
+    public function optOutJS(): string
```
I guess this new action is meant to be directly included externally. In that case this may also need some additional checks here:
https://github.com/matomo-org/matomo/blob/74e0ace978ad62c879abc736cdd1fca3894db5d3/core/FrontController.php#L541-L544
and maybe there:
https://github.com/matomo-org/matomo/blob/34e226540e69ab4b6dd82a9024d72782e1ac058a/core/Session.php#L177

sgiehl commented on plugins/PrivacyManager/vue/src/OptOutCustomizer/OptOutCustomizer.vue:
```diff
@@ -183,23 +302,25 @@ export default defineComponent({
+        fontColor: this.fontColor.substr(1),
+        fontSize: this.fontSizeWithUnit,
+        fontFamily: this.fontFamily,
+        showIntro: (this.showIntro === true ? 1 : 0),
```
Currently if `show intro` is not selected, the resulting url contains `&showIntro=`, might be better if it would contain `&showIntro=0` maybe, as otherwise it might look like the parameter could be simply removed, which isn't the case. Guess the values here needs to be strings to achieve that...

sgiehl commented on plugins/CoreAdminHome/OptOutManager.php:
```diff
@@ -248,13 +613,13 @@ private function optOutStyling()
         }
 
-        if ($cssfontfamily && (preg_match('/^[a-zA-Z0-9-\ ,\'"]+$/', $cssfontfamily))) {
+        if ($cssfontfamily && (preg_match('/^[a-zA-Z0-9-,\'"]+$/', $cssfontfamily))) {
```
any reason for disallowing spaces here? won't this break when sending a font family like `'Helvetica Neue'`

bx80 replied:
Good spot :+1: PHPStorm incorrectly flagged this as a redundant escape character and I naively applied the adjustment :man_facepalming:. I'll add a `@noinspection RegExpRedundantEscape` annotation to suppress this.

sgiehl commented on plugins/CoreAdminHome/OptOutManager.php:
```diff
@@ -248,13 +613,13 @@ private function optOutStyling()
         }
 
-        if ($cssfontsize && (preg_match("/^[0-9]+[\.]?[0-9]*(px|pt|em|rem|%)$/", $cssfontsize))) {
+        if ($cssfontsize && (preg_match("/^[0-9]+[.]?[0-9]*(px|pt|em|rem|%)$/", $cssfontsize))) {
```
doesn't removing the `\` mean, that instead of an optional `.`, now any character can be there? 🤔

bx80 replied:
As above: `@noinspection RegExpRedundantEscape` added.

sgiehl reviewed (changes requested):
Had a look through the code and left a couple of comments. Also did some basic functional testing and so far everything was working as expected (at least what I would expect from it)

Commit:
Merge branch '4.x-dev' into m-17452-opt-out-first-party

Commit:
Show intro parameter show as zero if not set, hide opt out complete message if not showing intro, added additional opt out url exclusions, reverted incorrect regex adjustments

bx80 committed:
built vue files

Commit:
Merge branch '4.x-dev' into m-17452-opt-out-first-party

Commit:
Annotated new core home embed code methods as internal

sgiehl commented on plugins/CoreAdminHome/OptOutManager.php:
```diff
@@ -158,10 +181,333 @@ public function getQueryParameters()
+        }
+        
+        function checkForMatomoTracker() {
+            if (typeof _paq === 'undefined') {                
+                if (checkForTrackerTried < checkForTrackerTries) {
+                    setTimeout(checkForMatomoTracker, checkForTrackerInterval);
+                    checkForTrackerTried++;
+                    return;
+                } else {
+                    if (settings.useCookiesIfNoTracker) {
+                        showOptOutDirect();
+                        return;
+                    } else {
+                        console.log('Matomo OptOutJS: failed to find Matomo tracker after '+(checkForTrackerTries*checkForTrackerInterval / 1000)+' seconds');
+                        return;    
+                    }
+                }
+            }
+            showOptOutTracker();
```
Guess this could be written a bit simpler like:
```suggestion
            if (typeof _paq !== 'undefined') {
                showOptOutTracker();
                return;
            }
            
            if (checkForTrackerTried < checkForTrackerTries) {
                setTimeout(checkForMatomoTracker, checkForTrackerInterval);
                checkForTrackerTried++;
                return;
            }
            
            if (settings.useCookiesIfNoTracker) {
                showOptOutDirect();
                return;
            }
            
            console.log('Matomo OptOutJS: failed to find Matomo tracker after '+(checkForTrackerTries*checkForTrackerInterval / 1000)+' seconds');
```

sgiehl commented on plugins/CoreAdminHome/API.php:
```diff
@@ -335,4 +342,49 @@ private function getDatesToInvalidateFromString($dates, $period)
+     * @param string $fontSize
+     * @param string $fontFamily
+     * @param bool $showIntro
+     * @param string $piwikUrl
```
could you rename that parameter to `matomoUrl` everywhere, where it won't break BC?

sgiehl commented on plugins/CoreAdminHome/OptOutManager.php:
```diff
@@ -158,10 +181,333 @@ public function getQueryParameters()
+                                         string $fontSize, string $fontFamily, bool $showIntro): string
+    {
+        return '<div id="matomo-opt-out"></div>
+<script src="'.$piwikUrl.'/index.php?module=CoreAdminHome&action=optOutJS&div=matomo-opt-out&language='.$language.'&backgroundColor='.$backgroundColor.'&fontColor='.$fontColor.'&fontSize='.$fontSize.'&fontFamily='.$fontFamily.'&showIntro='.($showIntro ? '1' : '0').'"></script>';
```
Here you could apply a `rtrim($piwikUrl, '/')` to ensure the url only contains one `/`

sgiehl reviewed (changes requested):
Left some other comments for improvements.
Besides that I was wondering if it might make sense to also add a select box for language selection. There might be cases were someone might use Matomo in another language he wants to provide the opt out in.
While this might not be an issue when the opt out is loaded via a script tag, this would be an issue if he decides to include the JS, as the code already contains all translations.
So guess we should at least provide this option when the js code is embedded...

bx80 committed:
Simplify code

Co-authored-by: Stefan Giehl <stefan@matomo.org>

Commit:
Updated parameter name, added trailing url trimming

Commit:
Added language choice selection for self-contained opt-out code

Commit:
Increased test_TotalPiwikFilesSize_isWithinReasonnableSize size from 58MiB to 60MiB

Commit:
Adjust layout

Commit:
Update UI test screenshot

sgiehl commented on plugins/CoreAdminHome/Controller.php:
```diff
@@ -255,11 +255,22 @@ public function trackingCodeGenerator()
+     */
+    public function optOutJS(): string
+    {
+        return $this->optOutManager->getOptOutJS();
```
This method currently seems to start the session. This shouldn't be required as the method only serves JS code.
Guess we can avoid the session start by adding an additional check here:
https://github.com/matomo-org/matomo/blob/df0fb6290ce496b9520e2062293aeb32c68978af/core/FrontController.php#L705-L709

bx80 replied:
Good spot :+1: I've added an exclusion to the front controller.

sgiehl replied:
Works 👍  Another thing that just came to my mind was, if we should send some additional headers here. Like the correct content type or some caching related headers. 🤔

bx80 replied:
That makes sense, caching would prevent any server side changes being applied via the inline settings data :+1: I've added a content type header for 'application/javascript' and set caching to 'no-store'.

sgiehl reviewed (changes requested):
Besides the comment I have left, this looks good to me so far.
But might be good if maybe @Findus23 and/or @tsteur could also have a look at that one. As the OptOut is quite important, we should avoid producing any regressions there...

Commit:
Prevent session start for optOutJS controller method

Commit:
Merge branch '4.x-dev' into m-17452-opt-out-first-party

Commit:
Fix conflicts

Commit:
Fix for vue error

bx80 committed:
built vue files

Commit:
Added content type and cache control headers when serving the opt out javascript

Commit:
Merge remote-tracking branch 'origin/m-17452-opt-out-first-party' into m-17452-opt-out-first-party

tsteur commented:
👍  had a quick look

## UI 
Regarding the opt out screen:

<img width="1289" alt="image" src="https://user-images.githubusercontent.com/273120/187125728-af8a3cb3-65a0-49cd-bade-1821bd32f34b.png">

I wonder if it makes sense to add a checkbox whether to apply any custom styling or not? I would assume by default no custom styles are needed (which is different now to before) as the content is rendered inside the regular page and thus automatically gets the right style. Vs by applying the styles by default we need to configure them correctly and adjust them.  I mean the `font color, backgroundColor, fontSize and fontFamily`. Maybe these we could not apply by default and "hide" them behind a setting? Not sure if it's clear what I mean and if it makes sense?

The `Remember to test your opt-out` is great. I wonder if we can make it slightly more visible using a blue info background or position it somewhere else or so.

Maybe this last part on the opt out page we could give it's own title as otherwise it's barely visible?
<img width="847" alt="image" src="https://user-images.githubusercontent.com/273120/187126199-e60f8ac4-8b0e-493a-a3d1-bd97a4fa7277.png">

## Opt Out

It wasn't fully clear what to test all. I wonder if it makes sense to document the behaviour of the opt out for all the different cases so it's clear for everyone how we expect the JS opt out to work?

FYI also added few samples to https://github.com/matomo-org/test-examples/pull/13 for easier testing so that not everybody needs to create the same scripts. Not sure what else is missing. I assume custom cookie configuration etc.

Is maybe also a "What's new" entry missing? Basically, people will need to become aware that the new optOut should be used and the older one no longer. Might also need to discuss with marketing to promote this in newsletter?

sgiehl commented:
> Is maybe also a "What's new" entry missing? Basically, people will need to become aware that the new optOut should be used and the older one no longer. Might also need to discuss with marketing to promote this in newsletter?

Guess an entry in the changelog also makes sense. As including the opt out in most cases is development work, might be useful to have it in there.

Commit:
Added option to apply styling (disabled by default), improved reminder box visibility, updated developer log, added what's new, added title to opt-out developer guide section

justinvelluppillai commented on plugins/PrivacyManager/changes.json:
```diff
@@ -0,0 +1,7 @@
+[
+  {
+    "title":       "Privacy Manager - User Opt-Out Improvements",
+    "description": "The privacy manager user opt-out has been updated to provide new code generation options that improve compatibily and allow additional customisation.",
```
```suggestion
    "description": "The privacy manager user opt-out has been updated to provide new code generation options that improve compatibility and allow additional customisation.",
```

bx80 committed:
built vue files

bx80 committed:
Fix typo

Co-authored-by: Justin Velluppillai <justin@innocraft.com>

Commit:
Update tests

Commit:
Merge remote-tracking branch 'origin/m-17452-opt-out-first-party' into m-17452-opt-out-first-party

Commit:
Merge branch '4.x-dev' into m-17452-opt-out-first-party

Commit:
Update UI test screenshot

bx80 commented:
Thanks for the feedback @tsteur and @sgiehl 

I've made the following changes:
- Added a checkbox to toggle custom styling
- Made the 'Remember to test...' box more visible (bootstrap blue background)
- Added a separate header for the developer guide for custom opt-out section: 'Build your own'
- Updated the change log to explain the new options and that the iframe call will still work
- Added a 'What's new?' entry

The code should now be ready for further review.

To document the various ways the opt-out can work, would it be better to add another section to this guide: https://developer.matomo.org/guides/tracking-javascript-guide#optional-creating-a-custom-opt-out-form or maybe the custom opt-out form guide and the new information about how the opt-out can work should be moved to a separate developer docs section?

I can create draft updates of both these FAQs too:
https://matomo.org/faq/general/faq_20000/
https://matomo.org/faq/general/configure-privacy-settings-in-matomo/
Not sure if there is any other documentation to update?

sgiehl reviewed (commented):
Had a look through the code changes. Generally looks fine, but we should try to avoid adding `style` attributes to elements where possible. Otherwise it won't be easily possible for themes to changes those elements. Would be better to either reuse existing css classes if possible or to add new classes if needed.

Commit:
Moved inline CSS style to less

bx80 committed:
built vue files

Commit:
Update UI test screenshot

Commit:
Merge remote-tracking branch 'origin/m-17452-opt-out-first-party' into m-17452-opt-out-first-party

sgiehl commented on plugins/PrivacyManager/changes.json:
```diff
@@ -0,0 +1,7 @@
+  {
+    "title":       "Privacy Manager - User Opt-Out Improvements",
+    "description": "The privacy manager user opt-out has been updated to provide new code generation options that improve compatibility and allow additional customisation.",
+    "version":     "4.12.0"
```
If we plan to write some kind of documentation around the new opt out, we could include the link here.

sgiehl reviewed (approved):
Left a minor comment, we can also apply after hat one has been merged.
Otherwise this now looks good. So if @tsteur doesn't have any further objections, this should be good to merge.

tsteur commented:
Great to see our first what's new entry 🎉  That's awesome.

Had another quick look. I believe the information box uses a custom color that we maybe usually don't use in that combination? 
<img width="734" alt="image" src="https://user-images.githubusercontent.com/273120/188355129-68f47640-df2d-4878-a685-7d66cf76c5e3.png">

Using class `system notification notification-info` + white color should do the trick.

<img width="861" alt="image" src="https://user-images.githubusercontent.com/273120/188355103-db995bcc-f573-441d-b259-dd657c6ba50f.png">

I haven't tested the opt out otherwise. It be great to only merge though once we've created all the documentation (especially all the different flows and what we expect to happen in which cases), and possibly a small blog post? Not sure what's there already. Just wanting to make sure we won't forget it.

Commit:
Merge branch '4.x-dev' into m-17452-opt-out-first-party

Commit:
Update test reminder section style

bx80 committed:
built vue files

Commit:
Update UI test screenshot

Commit:
Merge remote-tracking branch 'origin/m-17452-opt-out-first-party' into m-17452-opt-out-first-party

Commit:
Accept cookie path and cookie domain as URL parameters, revert loading (incorrect) cookie settings from config

Commit:
Bump to retry random test fail

Commit:
Merge branch '4.x-dev' into m-17452-opt-out-first-party

Commit:
Local vue build

bx80 committed:
built vue files

Commit:
Added blog post link to the what's new notification

Commit:
Merge remote-tracking branch 'origin/m-17452-opt-out-first-party' into m-17452-opt-out-first-party

Commit:
Update UI test screenshot

Commit:
Update "What's New?" link url
